### PR TITLE
tofieldset: Add tests to show it already allow duplicates

### PR DESCRIPTION
**Pull request by apelisse**
Continuation of #247.

This uses the pattern that I've generally used to solve the problem, which is to not try to be too smart about duplicate entries, and just record the path element for the whole thing, rather than try to track all the fields within. It's not super relevant because no one should ever be able to apply duplicates, so they wouldn't really have conflicts with sub-fields anyway, we've decided that they should have a conflict with the idea of "duplicates" (de-duplicating is a conflict).

/assign @alexzielenski

### CODE REVIEW

File: typed/toset_test.go [hunk 1]
```diff
@@ -244,6 +249,16 @@ var fieldsetCases = []fieldsetTestCase{{
 			_P("list", _KBF("key", "b", "id", 1), "key"),
 			_P("list", _KBF("key", "b", "id", 1), "id"),
 		)},
+		{`{"list":[{"key":"a","id":1,"nv":2},{"key":"a","id":2,"nv":3},{"key":"b","id":1},{"key":"a","id":2,"bv":true}]}`, _NS(
+			_P("list", _KBF("key", "a", "id", 1)),
+			_P("list", _KBF("key", "a", "id", 1), "key"),
+			_P("list", _KBF("key", "a", "id", 1), "id"),
+			_P("list", _KBF("key", "a", "id", 1), "nv"),
+			_P("list", _KBF("key", "a", "id", 2)),
```
Comment (apelisse) [inline typed/toset_test.go:257]: cc @jpbetz 
This code is mostly irrelevant because we only use `ToFieldSet` on objects that are applied, and we should never allow apply to have duplicates (should we change that code?). I still wanted to illustrate that when someone will update with duplicates items, we will just give them ownership of the associative key and that's it, as a way to indicate that they own that key multiple times. WDYT?

File: typed/tofieldset.go [hunk 2]
```diff
@@ -94,9 +94,31 @@ func (v *toFieldSetWalker) doScalar(t *schema.Scalar) ValidationErrors {
 }
 
 func (v *toFieldSetWalker) visitListItems(t *schema.List, list value.List) (errs ValidationErrors) {
+	// Keeps track of the PEs we've seen
+	seen := fieldpath.MakePathElementSet(list.Length())
+	// Keeps tracks of the PEs we've counted as duplicates
+	duplicates := fieldpath.MakePathElementSet(list.Length())
```
Comment (alexzielenski) [inline typed/tofieldset.go:100]: I would've expected the implementation to use just one set and one loop; `if seen.Has(pe) { continue }`

Comment (apelisse) [reply typed/tofieldset.go:100]: we're trying to add duplicates exactly once to the set. I need to know if I've already seen them to know if they are duplicates, and I need to know if I've already added them to the set.

**Comment by apelisse** [comment]
/hold

I'm trying a new test with pruning fields that are removed by appliers and this doesn't work, looking into this.

File: typed/tofieldset.go (hunk 2 above)
Comment (alexzielenski) [reply typed/tofieldset.go:100]: I guess Im proposing the following pseudocode: 

```go
seen := NewSet()
for item in list {
    if seen.Has(item) { continue }
    seen.Insert(item)
    listWithoutDuplicates.append(item)
}
```

Doesn't this do as you described?

Comment (apelisse) [reply typed/tofieldset.go:100]: Yeah, I think that makes sense, thanks.

File: typed/toset_test.go (hunk 1 above)
Comment (apelisse) [reply typed/toset_test.go:257]: One drawback of this, but I don't know if we can solve it anyway, is that if someone applies one item, and then it gets updated into two items, that item won't be pruned when it's removed from the applied configuration, meaning it can no longer be removed with server-side apply. I think that's reasonable.

**Comment by apelisse** [comment]
/hold cancel

Thought about this further, I think it's fine, unless I forgot the use-case I had in mind ...

File: typed/toset_test.go (hunk 1 above)
Comment (alexzielenski) [reply typed/toset_test.go:257]: > This code is mostly irrelevant because we only use ToFieldSet on objects that are applied, and we should never allow apply to have duplicates (now, that being said, I'm curious if this whole change even makes sense?)

If apply can never have duplicates why are we making SMD tolerate duplicates?

Comment (alexzielenski) [inline typed/toset_test.go:257]: > if someone applies one item, and then it gets updated into two items, that item won't be pruned when it's removed from the applied configuration, meaning it can no longer be removed with server-side apply.

Why wouldn't it be pruned?

File: typed/tofieldset.go (hunk 2 above)
Comment (alexzielenski) [reply typed/tofieldset.go:100]: can mark this as resolved, since after discussion offline it is not what you want

**Review by alexzielenski — COMMENTED** [review]
Just to clarify, this change makes SMD give ownership of the entire object if there are duplicates in a list map?

I'm having a hard time tracking the "before" behavior and the "after" behavior.

File: typed/toset_test.go (hunk 1 above)
Comment (apelisse) [reply typed/toset_test.go:257]: > If apply can never have duplicates why are we making SMD ToFieldSet tolerate duplicates?

Context, we're trying to solve #234

For built-ins, the `listType=map` is used a little loosely, and we don't have a great way to address that, except for accepting some duplicates. We still can't apply duplicates, but right now you can't apply an objects that has these duplicates.

> How do Updates get the list of fields to own if not via ToFieldSet?

By using "Compare", they own the fields that they changed, since we don't know exactly what they intended to own.

Comment (apelisse) [reply typed/toset_test.go:257]: Because when you update to multiple items, it steals the ownership of the whole item and we don't prune things that you don't have ownership over.

Comment (alexzielenski) [reply typed/toset_test.go:257]: > We still can't apply duplicates, but right now you can't apply an objects that has these duplicates.

Is ToFieldSet invoked on the final merged object, or the partial patch? If it is invoked on the partial patch aren't there no duplicates and this PR has no effect?

Comment (apelisse) [reply typed/toset_test.go:257]: Correct, that's more or less what I said in the PR body:
> It's not super relevant because no one should ever be able to apply duplicates

But:
1. This is for consistency, if we agree that this is how ownership of things that are duplicates, we should probably apply it everywhere, even if it's not relevant here.
2. This is the first time we use that new pattern in the code so I want to make sure we're fine with it.

Comment (alexzielenski) [reply typed/toset_test.go:257]: That seems like correct behavior to me. What if it was a map typed list with duplicate keys but unequal objects?

Apply, "applier": [ObjWithKeyA]
Update, "updater": [ObjWithKeyAButSlighlyDifferent, ObjWithKeyAButAlsoDifferent] # Should be three copies of 'A'?
Apply, "applier": [] # is A pruned?

Comment (apelisse) [reply typed/toset_test.go:257]: Very good point.

So first of all, `Update` replaces the whole list, so if you do
> Update, "updater": [ObjWithKeyAButSlighlyDifferent, ObjWithKeyAButAlsoDifferent] # Should be three copies of 'A'?

you end-up with that exact object.

Second, the way this is meant to work, is that if you have duplicate, you own the key and the key only, so here's what happens in your scenario:

Apply, "applier": [ObjWithKeyA] -> applier owns ObjWithKeyA
Update, "updater": [ObjWithKeyAButSlighlyDifferent, ObjWithKeyAButAlsoDifferent] -> duplicates, updater owns just "KeyA" (none of the fields)
Apply, "applier": [] # is A pruned? no
Whats the final object
[ObjWithKeyAButSlighlyDifferent, ObjWithKeyAButAlsoDifferent]  with KeyA owned by updater.

Comment (alexzielenski) [reply typed/toset_test.go:257]: i think thats reasonable

**Comment by alexzielenski** [comment]
LGTM from me. I think you mentioned you wanted @jpbetz input so will defer tag

**Comment by alexzielenski** [comment]
whoops
